### ISSUE TITLE
${Callsite} method name for async methods are not correct in .NET Core

### ISSUE THREAD

**Issue by 304NotModified** [NLog/NLog#1805]
Partly moved from https://github.com/NLog/NLog.Extensions.Logging/issues/67

Current:` Auth.Controllers.LoginController+<Post>d__4.MoveNext`
should be: `Auth.Controllers.LoginController.Post`

`callsite:CleanNamesOfAnonymousDelegates` won't change anything.

With layout: `layout="${longdate}|${uppercase:${level}}|${callsite:className=true:methodName=true}|${message}${newline}${stacktrace}"`

Using 
`${callsite:CleanNamesOfAnonymousDelegates=true:ClassName=false}
`I get
```
2016-11-29 11:53:39.2654|INFO|Logged In foo@bar.com
MoveNext
ExecutionContext.Run => ExecutionContext.RunInternal => <Post>d__4.MoveNext

```
of course I'm looking to get something like Auth.LoginController.Post (fully qualified class name + method name)

using 
`${callsite:CleanNamesOfAnonymousDelegates=true:MethodName=false}`
I get 
```
2016-11-29 11:54:58.6905|INFO|Logged In foo@bar.com
Auth.Controllers.LoginController+<Post>d__4
ExecutionContext.Run => ExecutionContext.RunInternal => <Post>d__4.MoveNext

```

Current code: 
https://github.com/NLog/NLog/blob/e77dba48c9140caeb8c0a28a54f92a3ee93b82a0/src/NLog/LayoutRenderers/CallSiteLayoutRenderer.cs#L145-L185

**Comment by 304NotModified** [comment]
in methodname it searches for `<`, but it seems that methodname and classname are mixed.

**Comment by 304NotModified** [comment]
@EltonInAtlanta could you provide us a sample call (including the method definition)? I would be nice if the case is without DI

**Comment by EltonInAtlanta** [comment]
Here is the controller (sanitized slightly) I've been using throughout this discussion.

```
    public class LoginController :Controller
    {
        private readonly TokenProviderOptions TokenProviderOptions;
        private readonly ILogger Logger;
        private readonly IMembership Membership;

        public LoginController(TokenProviderOptions options, ILoggerFactory logFactory, IMembership membership)
        {
            TokenProviderOptions = options;
            Logger = logFactory.CreateLogger<LoginController>();
            Membership = membership;
        }

        [HttpPost]
        [AllowAnonymous]
        public async Task<IActionResult> Post([FromBody] Credentials creds)
        {
            if (creds != null)
            {
                if (creds.Application == null) creds.Application = MembershipApplications.ApplicationPublicNameParent;

...
                if (user == null)
                {
                    Logger.LogInformation($"Failed LogIn {creds.Username}");
                    return new UnauthorizedResult();
                }

...

                Logger.LogInformation($"Logged In {creds.Username}");
                return new JsonResult(token);
            }
            else
            {
                return new BadRequestResult();
            }

        }
    }

```

Here's the relevant part of the startup configuration:
```
        public void Configure(IApplicationBuilder app, IHostingEnvironment env, ILoggerFactory loggerFactory)
        {
...            
                loggerFactory.WithFilter(
                    new FilterLoggerSettings
                    {
                        { "Microsoft", LogLevel.Warning },
                        { "System", LogLevel.Warning }
                    }
                    ).AddNLog(); 
                env.ConfigureNLog(@".\nlog.config");

                NLog.LogManager.AddHiddenAssembly(Assembly.Load(new AssemblyName("Microsoft.Extensions.Logging.Filter")));

                ILogger log = loggerFactory.CreateLogger("Configure");
                log.LogInformation("Startup");
                DatabaseWarmupRead();
                log.LogInformation("Warmed Up");
               
  ```

**Comment by EltonInAtlanta** [comment]
For comparison, I added a non DI created logger to test:
```
    public class AnnonTestController : Controller
    {
        // GET api/values/5
        [HttpGet("{id}")]
        public string Get(int id)
        {
            NLog.Logger logger = NLog.LogManager.GetLogger("AnnonTest");
            logger.Info("How now brown cow");

            return "value";
        }

```
Here is the output from the login call in the previous post, and the call to this test controller:
```
016-12-01 10:40:03.6020|INFO|How now brown cow
Auth.Controllers.AnnonTestController.Get
<InvokeActionMethodAsync>d__27.MoveNext => <no type>.lambda_method => AnnonTestController.Get

2016-12-01 10:41:14.5523|INFO|Logged In foo@bar.com
Auth.Controllers.LoginController+<Post>d__4.MoveNext
ExecutionContext.Run => ExecutionContext.RunInternal => <Post>d__4.MoveNext

```
This is the NLog configuration producing this output:
```
    <target xsi:type="File" name="debugFile" fileName="${basedir}/logs/debug-${shortdate}.log"
            layout="${longdate}|${uppercase:${level}}|${message}${newline}${callsite:CleanNamesOfAnonymousDelegates=true}${newline}${stacktrace}${newline}${newline}" />

```

**Comment by 304NotModified** [comment]
> For comparison, I added a non DI created logger to test:

It's even a problem when the action isn't `async`?

**Comment by EltonInAtlanta** [comment]
Issue appears to be async only.

Here's my modified test controller:
```
using System.Collections.Generic;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.Authorization;
using Auth.Code;
using Microsoft.Extensions.Logging;
//using wallet.Code;

// For more information on enabling Web API for empty projects, visit http://go.microsoft.com/fwlink/?LinkID=397860

namespace Auth.Controllers
{
    [Route("api/[controller]")]
    [AllowAnonymous]
    public class AnnonTestController : Controller
    {

        private readonly ILogger Logger;

        public AnnonTestController(ILoggerFactory logFactory)
        {
            Logger = logFactory.CreateLogger<AnnonTestController>();
        }

        [HttpGet]
        public async Task<IEnumerable<string>> Get()
        {
            NLog.Logger logger = NLog.LogManager.GetLogger("AnnonTest");
            logger.Info("Native, async method");

            Logger.LogInformation("Injected, async method");

            return await Task.FromResult( new string[] { "value1", "value2" });
        }

        // GET api/values/5
        [HttpGet("{id}")]
        public string Get(int id)
        {
            NLog.Logger logger = NLog.LogManager.GetLogger("AnnonTest");
            logger.Info("Native, not async");

            Logger.LogInformation("Injected, not async");
            return "value";
        }
}
}

```
And the output:
```
2016-12-01 16:53:45.4233|INFO|Startup
Auth.Startup.Configure
RuntimeMethodInfo.UnsafeInvokeInternal => RuntimeMethodHandle.InvokeMethod => Startup.Configure

2016-12-01 16:53:51.2155|INFO|Warmed Up
Auth.Startup.Configure
RuntimeMethodInfo.UnsafeInvokeInternal => RuntimeMethodHandle.InvokeMethod => Startup.Configure

2016-12-01 16:53:54.8998|INFO|Native, not async
Auth.Controllers.AnnonTestController.Get
<InvokeActionMethodAsync>d__27.MoveNext => <no type>.lambda_method => AnnonTestController.Get

2016-12-01 16:53:54.9627|INFO|Injected, not async
Auth.Controllers.AnnonTestController.Get
<InvokeActionMethodAsync>d__27.MoveNext => <no type>.lambda_method => AnnonTestController.Get

2016-12-01 16:53:58.4663|INFO|Native, async method
Auth.Controllers.AnnonTestController.Get
<InvokeActionMethodAsync>d__27.MoveNext => <no type>.lambda_method => AnnonTestController.Get

2016-12-01 16:53:58.4975|INFO|Injected, async method
Auth.Controllers.AnnonTestController.Get
<InvokeActionMethodAsync>d__27.MoveNext => <no type>.lambda_method => AnnonTestController.Get

2016-12-01 16:54:28.5551|INFO|Logged In esaulsberry@horizonsoftware.com
Auth.Controllers.LoginController+<Post>d__4.MoveNext
ExecutionContext.Run => ExecutionContext.RunInternal => <Post>d__4.MoveNext
```
Looks pretty good here no matter what I do.  So now I'm trying to understand why that last one is special.  It's a call to the login controller in my first post on this thread.

**Comment by valuerr** [comment]
Gotted same bug on Nlog versions: `4.3.11 (stable)` and `5.0.0-beta03`

The code:

```
using System.Threading;
using System.Threading.Tasks;
using NLog;

namespace ConsoleApplicationPR
{
    class Program
    {
        private static Logger Log;

        private static async Task DoSomeWorkAsync()
        {
            Log.Debug("DoSomeWorkAsync:start");
            await Task.Delay(100);
            Log.Debug("DoSomeWorkAsync:exit");
        }

        static void Main(string[] args)
        {
            Log = LogManager.GetCurrentClassLogger();
            Log.Debug("Main:start");
            DoSomeWorkAsync();
            Log.Debug("Main:exit");
            Thread.Sleep(3000);
        }
    }
}
```

outputs:

```
d:/develop/workspaces/dotnet/ConsoleApplicationPR/ConsoleApplicationPR/bin/Debug/ConsoleApplicationPR.exe
02:11:34.8423 [DEBUG]  [1] (Program.cs:21) Main:start
02:11:34.8423 [DEBUG]  [1]  DoSomeWorkAsync:start
02:11:34.8643 [DEBUG]  [1] (Program.cs:23) Main:exit
02:11:34.9744 [DEBUG]  [7] (Program.cs:15) DoSomeWorkAsync:exit
```


simplified App.config:

```
<?xml version="1.0" encoding="utf-8"?>

<configuration>
  <configSections>
    <section name="nlog" type="NLog.Config.ConfigSectionHandler, NLog" />
  </configSections>

  <nlog xmlns="http://www.nlog-project.org/schemas/NLog.xsd"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="http://www.nlog-project.org/schemas/NLog.xsd NLog.xsd"
        autoReload="true"
        throwExceptions="false"
        internalLogLevel="Off" internalLogFile="c:\temp\nlog-internal.log">

    <variable name="callsiteFormat"
              value="${callsite:className=False:fileName=True:includeSourcePath=False:methodName=False:cleanNamesOfAnonymousDelegates=False:skipFrames=0}" />
    <!--    Not works with async methods; -->
    <!--    Todo: commit issue-->

    <variable name="NormalLayout"
              value="${time} [${level:uppercase=true}] [${threadid}] ${callsiteFormat} ${message}" />

    <targets>
      <target xsi:type="ColoredConsole"
              name="console"
              layout="${NormalLayout}" />
    </targets>

    <rules>
      <logger name="*" minlevel="Trace" writeTo="console" />
    </rules>
  </nlog>
</configuration>

```

**Comment by 304NotModified** [comment]
Thanks for the info !

Could you provide an unit test? That would boost the fix!

**Comment by valuerr** [comment]
Done.

P.S. The bug is actual on `msvc 2015, 14.0.25431.01 update 3`, `.NET 4.6.01586`

Didn't test it on Mono.

**Comment by 304NotModified** [comment]
@valuerr altough it's the same component, it's a different bug. I move your comments to a new issue, so it's easier to track/solve: https://github.com/NLog/NLog/issues/1813

**Comment by 304NotModified** [comment]
This issue (method name), seems only an issue in .NET Core, as I could not reproduce it in NLog 4.3 (cf0d7cfa25ea7a3104a766a81c26f9e875ee6761)

**Comment by tkhaugen** [comment]
The way async methods are implemented by the compiler, all code after an await-call to another async method is continued by a state machine in such a way that the call stack no longer leads back to the user code. The "skip a frame" approach implemented in LoggerImpl doesn't work in these cases.

I have humbly submitted pull request (#2281) which takes another approach, inspecting the stackframe in the Append-methods of CallSiteLyoutRenderer class. The state machine class generated by the compiler is nested in the user type, and so this type can be retrieved from the DeclaringType method of the state machine TypeInfo.

One of the unit tests included with the pr reproduces the problem as described here.

**Comment by jlechem** [comment]
I get this exact same issue in an asp.net web api 2 application using NLog 4.6.7 any async method has extraneous stuff in it. 

IvrRelayService.Controllers.CpasController+<CpasSelect>d__3.MoveNext

Non async methods are not like this

IvrRelayService.Controllers.CpasController.BuildEncodedGetUrl

**Comment by snakefoot** [comment]
@jlechem Thank you for reporting. Have created PR  #3641

**Comment by snakefoot** [comment]
@jlechem Thank you for reporting. Have created PR  #3641

What happens if you configure `${callsite:CleanNamesOfAsyncContinuations=true}` ?

**Comment by 304NotModified** [comment]
Please open a new issue and refer to this one. Thanks!

**Comment by jlechem** [comment]
@snakefoot Thanks for mentioning this I tried using that flag as per the renderer/layout documentation and it made no difference.

**Comment by jlechem** [comment]
Also I just realized this was a .net core original issue I AM NOT using .net core. This was noticed in a WebAPI 2.1 project. Hopefully I didn't screw anybody/anything up reporting it here.

**Comment by snakefoot** [comment]
@jlechem Think we should stop commenting in this issue, and you should create a new issue as just said. Remember to fill out the template and also provide details of what version of NetFramework in use.

**Comment by 304NotModified** [comment]
Please don't start new bug reports on closed GitHub issues. 

We need more (context) info and it's hard to follow closed issues. So please open a new issue, fill in the template and refer to this one. Thanks!
